test/com/google/javascript/jscomp/Es6VariableReferenceCheckTest.java

```diff
@@ -277,6 +277,10 @@ public void testClass() {
     assertNoWarning("class A { f() { return 1729; } }");
   }
 
+  public void testRedeclareClassName() {
+    assertNoWarning("var Clazz = class Foo {}; var Foo = 3;");
```

Comment:
I agree with @dimvar that we should ban (or at least discourage) this pattern, but I also think we should support it since third party code is already doing this.